src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -6,68 +6,128 @@ import is.hail.utils._
+
   val ordering: Ordering[Annotation] = pkType.ordering.toOrdering
-  require(rangeBounds.isEmpty || rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => ordering.compare(left, right) < 0 })
 
   def region: Region = rangeBounds.region
 
-  def loadElement(i: Int): Long = rangeBoundsType.loadElement(rangeBounds.region, rangeBounds.aoff, rangeBounds.length, i)
+  def loadElement(i: Int): Long = rangeBoundsType.loadElement(region, rangeBounds.aoff, rangeBounds.length, i)
+
+  def loadStart(i: Int): Long = pkIntType.loadStart(region, loadElement(i))
+
+  def loadEnd(i: Int): Long = pkIntType.loadStart(region, loadElement(i))
 
-  // return the smallest partition for which key <= max
   // pk: Annotation[pkType]
   def getPartitionPK(pk: Any): Int = {
     assert(pkType.typeCheck(pk))
-
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          ordering.compare(pk, rangeBounds(i)))
-    part
+    val part = rangeTree.queryValues(pkType.ordering, pk)
+    part.length match {
+      case 0 =>
+        if (pkType.ordering.gt(pk, rangeTree.root.get.maximum))
+          numPartitions - 1
+        else {
+          assert(pkType.ordering.lt(pk, rangeTree.root.get.minimum))
+          0
+        }
+      case 1 => part.head
+    }
   }
 
-  // return the smallest partition for which key <= max
-  // key: RegionValue[kType]
+  // return the partition containing pk
+  // needs to update the bounds
+  // key: RegionValue
   def getPartition(key: Any): Int = {
     val keyrv = key.asInstanceOf[RegionValue]
+    val keyUR = new UnsafeRow(kType, keyrv)
 
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          -pkKOrd.compare(rangeBounds.region, loadElement(i), keyrv))
-    part
+    val part = rangeTree.queryValues(pkKOrd, keyUR)
+
+    part.length match {
+      case 0 =>
+        if (pkKOrd.lt(rangeTree.root.get.maximum, keyUR))
+          numPartitions - 1
+        else {
+          assert(pkKOrd.gt(rangeTree.root.get.minimum, keyUR))
+          0
+        }
+      case 1 => part.head
+    }
   }
 
   def withKType(newPartitionKey: Array[String], newKType: TStruct): OrderedRVDPartitioner = {
-    val newPart = new OrderedRVDPartitioner(numPartitions, newPartitionKey, newKType, rangeBounds)
+    val newPart = new OrderedRVDPartitioner(newPartitionKey, newKType, rangeBounds)
     assert(newPart.pkType == pkType)
     newPart
   }
+
+  def copy(numPartitions: Int = numPartitions, partitionKey: Array[String] = partitionKey,
+    kType: TStruct = kType, rangeBounds: UnsafeIndexedSeq = rangeBounds): OrderedRVDPartitioner = {
+    new OrderedRVDPartitioner(partitionKey, kType, rangeBounds)
+  }
+
+  def remapRanges(newPartEnd: Array[Int]): OrderedRVDPartitioner = {
```

Comment:
I don't think this is remap, this is coalesce.  Are you coalescing the ranges or the partitions?  (Both?  Does it matter?)